projects/packages/my-jetpack/_inc/data/utils/get-my-jetpack-window-state.ts

```diff
@@ -1,17 +1,37 @@
-type MyJetpackState = Window[ 'myJetpackInitialState' ];
+type InitialState = Window[ 'myJetpackInitialState' ];
+type RestState = Window[ 'myJetpackRest' ];
+type ConnectionState = Window[ 'JP_CONNECTION_INITIAL_STATE' ];
 
-type ReturnType< A > = A extends undefined
-	? MyJetpackState
-	: MyJetpackState[ A extends keyof MyJetpackState ? A : never ];
+// Handles typing based on whether or not a key is provided. S = type of state, A = type of key
+type StateReturnType< S, A > = A extends undefined ? S : S[ A extends keyof S ? A : never ];
 
-const getMyJetpackWindowState = < A extends keyof MyJetpackState | undefined = undefined >(
+export const getMyJetpackWindowInitialState = <
+	A extends keyof InitialState | undefined = undefined,
+>(
 	key?: A
-): ReturnType< A > => {
+): StateReturnType< InitialState, A > => {
 	if ( ! key ) {
-		return window?.myJetpackInitialState as ReturnType< A >;
+		return window?.myJetpackInitialState as StateReturnType< InitialState, A >;
 	}
 
-	return ( window?.myJetpackInitialState?.[ key ] ?? null ) as ReturnType< A >;
+	return ( window?.myJetpackInitialState?.[ key ] ?? null ) as StateReturnType< InitialState, A >;
 };
 
-export default getMyJetpackWindowState;
+export const getMyJetpackWindowRestState = () => {
+	return ( window?.myJetpackRest ?? null ) as RestState;
```

Comment:
I just realized we cast potential null here... I think instead of null, we can return empty object - so that restructuring will not cause an error:

```suggestion
	return ( window?.myJetpackRest ?? {} ) as RestState;
```